Update the lewel extension

stefnotch opened the pull request:
Based on #1 

I'm sorry for this big PR. I can split it up into individual commits if it makes reviewing easier.

I did the following
- I created a new extension with the latest template https://code.visualstudio.com/api/get-started/your-first-extension . Now we depend on the latest VSCode tooling
  - Requires a newer VSCode version
  - Eslint is in the template
  - Comes with a unit test framework
  - Comes with launch configurations (just hit F5 and it starts up)
  - Adds a [`.vscodeignore`](https://code.visualstudio.com/api/working-with-extensions/publishing-extension#using-vscodeignore)
- Since I noticed that the other PR reformatted some code, I figured I'd add Prettier for code formatting. It's pretty much the most common code formatter in the JS world.
- Adds a complete textmate grammar for lelwel

The most important change is the complete textmate grammar. I basically wrote one from scratch by looking at the [neovim extension](https://github.com/0x2a-42/nvim-lelwel/blob/main/syntax/llw.vim), the lelwel parser itself and taking a few other grammars as a guideline for how to structure the rules.
It should now comprehensively cover every single token type, including semantic assertions[^1].

I, however, do not claim that this is perfect by any means. I'm pretty new to VSCode extension authoring.


[^1]: Semantic assertions seem to be missing in the nvim extension.

stefnotch commented:
@DevYatsu I'd love to get your opinions on this PR :)

DevYatsu commented:
Nice work, really appreciate the effort here!

Just a few suggestions:

- Adding test scaffolding was a great move! That said, it might be more helpful to include a minimal but meaningful test related to the project itself. The current test is just:
```javascript
	test('Sample test', () => {
		assert.strictEqual(-1, [1, 2, 3].indexOf(5));
		assert.strictEqual(-1, [1, 2, 3].indexOf(0));
	});
```
Maybe consider replacing it with something project-specific, even if it’s a placeholder.

- You could also consider adding `CONTRIBUTING.md` to `.vscodeignore` to keep the packaged extension lean.

- In `CONTRIBUTING.md`, it looks like the setup instructions are repeated, the strikethrough section at the beginning can probably be removed to avoid redundancy:

```
T̶h̶e̶n̶ ̶o̶p̶e̶n̶ ̶`̶s̶r̶c̶/̶e̶x̶t̶e̶n̶s̶i̶o̶n̶.̶t̶s̶`̶ ̶a̶n̶d̶ ̶p̶r̶e̶s̶s̶ ̶F̶5̶ ̶(̶o̶r̶ ̶s̶e̶l̶e̶c̶t̶ ̶*̶*̶R̶u̶n̶ ̶→̶ ̶S̶t̶a̶r̶t̶ ̶D̶e̶b̶u̶g̶g̶i̶n̶g̶*̶*̶)̶ ̶t̶o̶ ̶l̶a̶u̶n̶c̶h̶ ̶a̶ ̶n̶e̶w̶ ̶*̶*̶E̶x̶t̶e̶n̶s̶i̶o̶n̶ ̶D̶e̶v̶e̶l̶o̶p̶m̶e̶n̶t̶ ̶H̶o̶s̶t̶*̶*̶ ̶w̶i̶n̶d̶o̶w̶.̶

O̶p̶e̶n̶ ̶a̶ ̶.̶l̶l̶w̶ ̶f̶i̶l̶e̶ ̶i̶n̶ ̶t̶h̶a̶t̶ ̶w̶i̶n̶d̶o̶w̶ ̶t̶o̶ ̶t̶e̶s̶t̶ ̶t̶h̶e̶ ̶e̶x̶t̶e̶n̶s̶i̶o̶n̶

## Get up and running straight away

* Press `F5` to open a new window with your extension loaded.
* Run your command from the command palette by pressing (`Ctrl+Shift+P` or `Cmd+Shift+P` on Mac) and typing `Hello World`.
* Set breakpoints in your code inside `src/extension.ts` to debug your extension.
* Find output from your extension in the debug console.
```

Since the “Get up and running” section already explains this clearly, the earlier version can be trimmed.

Also, I hadn’t actually intended to reformat the code, so good call on adding Prettier.

stefnotch commented:
Thank you so much for the review! I'll definitely fix that :)

0x2a-42 commented:
Thanks a lot to both of you. I have now merged #1, so it will probably be easier to review this PR, if you rebase it first.

stefnotch commented:
This should now be ready for reviewing and merging :)

I'm looking at the whole "lelwel [with wasm](https://code.visualstudio.com/blogs/2024/06/07/wasm-part2#_a-webassembly-based-language-server)" and it seems reasonably straightforward.

Except for tokio's stdin and stdout [not working in a wasm environment](https://github.com/tokio-rs/tokio/issues/6516)
To run it in wasm, I'll have to replace [the stdin and stdout usage](https://github.com/0x2a-42/lelwel/blob/3f5b163f00fc3b9d98142ef2086ea73b94aedeb9/src/bin/lelwel-ls.rs#L10-L11) with a custom stream. I'll probably [do it like this](https://github.com/polarity-lang/polarity/blob/main/web/crates/browser/src/lib.rs#L51-L64).

0x2a-42 commented on .vscodeignore:
```diff
@@ -0,0 +1,12 @@
+src/**
+.gitignore
+.yarnrc
+vsc-extension-quickstart.md
```
This file from the template is not present, so we could remove it from the `.gitignore` file.

0x2a-42 commented on src/test/helper.ts:
```diff
@@ -0,0 +1,38 @@
+  const ext = vscode.extensions.getExtension("0x2a-42.lelwel")!;
+  await ext.activate();
+  doc = await vscode.workspace.openTextDocument(docUri);
+  editor = await vscode.window.showTextDocument(doc);
```
In the original file this code is surrounded by a `try` statement. Was there a reason to change this?

https://github.com/microsoft/vscode-extension-samples/blob/cf30922221d1c2dcb819da857b7f64583991fed0/lsp-sample/client/src/test/helper.ts#L21-L27

stefnotch replied:
Good question. I'm not sure why it is in a try statement, but I suppose imitating the original code makes sense?

0x2a-42 commented on src/extension.ts:
```diff
@@ -8,7 +8,7 @@ import {
 let client: LanguageClient;
 
-export function activate(_context: ExtensionContext) {
+export async function activate(_context: ExtensionContext) {
```
Why was adding `async` necessary? Apparently it's fine to make the `activate` function async, but I am just curious if there is any benefit for changing this.

stefnotch replied:
Good question! I made it async so that it only finishes after the language server has actually started. It lets me avoid doing the `sleep(2000)` in `helper.ts`.

And it's easier to test it in `extension.test.ts`.

```
async function testCompletion(
  docUri: vscode.Uri,
  position: vscode.Position,
  expectedCompletionList: vscode.CompletionList,
) {
  await activate(docUri); // waits for it to be started up
```

0x2a-42 commented on tsconfig.json:
```diff
@@ -1,13 +1,15 @@
+    /* Additional Checks */
+    // "noImplicitReturns": true, /* Report error when not all code paths in function return a value. */
+    // "noFallthroughCasesInSwitch": true, /* Report errors for fallthrough cases in switch statement. */
+    // "noUnusedParameters": true,  /* Report errors on unused parameters. */
```
I think we can remove those comments.

stefnotch replied:
Sounds good

0x2a-42 commented on .gitignore:
```diff
@@ -1,2 +1,3 @@
 /node_modules
 /out
+/.vscode-test
```
The template also adds a line `*.vsix` which could be useful to prevent accidental commits of VSIX package files.

stefnotch replied:
Oh, I must have missed that. I'll add it

0x2a-42 commented on package-lock.json:
Those are a lot of new dependencies, but I guess that's just how javascript works. :laughing:

I remember removing `eslint` and `mocha` when I wrote the original extension, because it seemed a bit excessive for what is essentially 30 lines of typescript. However, I think it will be fine, once the extension gets additional features, especially if multiple people commit to the repo. Also the new completion tests are really nice, as the language server was not tested before.

stefnotch replied:
I must admit I was also very torn on whether to add them or not.

And thank you! I must admit, I was pretty happy when I finally got that test to work.

0x2a-42 commented on CONTRIBUTING.md:
```diff
@@ -9,6 +9,24 @@ npm install
+## Get up and running straight away
 
-Open a .llw file in that window to test the extension.
+* Press `F5` to open a new window with your extension loaded.
+* Open a .llw file in that window to test the extension.
+* Set breakpoints in your code inside `src/extension.ts` to debug your extension.
+* Find output from your extension in the debug console.
```
We should replace "your extension" with "the extension".

stefnotch replied:
Good point, fixing

stefnotch commented:
I got very excited due to https://github.com/0x2a-42/lelwel/issues/30 and found time to take another stab at lelwel with wasm.

It now works! ✨ 
![image](https://github.com/user-attachments/assets/e1f0c920-5b56-4dd1-be03-b8451b263370)


So, in the latest commits, I added the following 
- A setting to toggle between lelwel wasm and native lelwel. It defaults to wasm.
- All the setup for running an extension in wasm
  - I followed the patterns from https://github.com/microsoft/vscode-extension-samples/tree/main/wasm-language-server
  - Notably, this means adding a `extensionDependencies` on `ms-vscode.wasm-wasi-core`. When the lelwel extension is installed, VSCode asks the user to also install that other extension. Hopefully Microsoft will make the wasm-wasi extension a built-in extension in the future.
- And I addressed the review comments

Hopefully looking through those commits to review the latest changes is reasonably doable.


My biggest question is: How should one generate the wasm binary for the VSCode extension?
Currently it's a manual process where one runs a command in the main `lelwel` repository, and then copies the output to the extension.

Alternatives that are more automated would be
- A script that does `git clone lelwel` at a specific commit ID and then runs the `cargo rustc` command. The downside there is that one needs to remember to update a commit ID in a Javascript file.
- Using git submodules. Though I try to avoid that whenever possible.
- Moving lelwel code around such that most of the lsp code is in a lelwel library. And then we'd create a Rust project in the vscode-lelwel `/server` folder that depends on the lelwel library.
- Moving `/vscode-lelwel` into the lelwel repository. Monorepos are very convenient.

0x2a-42 commented:
@stefnotch: Excellent work! Thanks a lot for your contribution.

> My biggest question is: How should one generate the wasm binary for the VSCode extension?

I thought about this as well and you probably listed all the reasonable solutions available. I'll try to use a submodule first, but maybe we switch to a monorepo later. `rust-analyzer` also seems to use a [monorepo](https://github.com/rust-lang/rust-analyzer/tree/master/editors/code).

stefnotch commented:
@0x2a-42 Sounds good! I look forward to being able to use this extension for development.

It's so nice when tools *just work* for my collaborators
